library/rtsp/src/main/java/com/google/android/exoplayer2/source/rtsp/media/MediaSession.java

```diff
@@ -0,0 +1,475 @@
+        // Do nothing.
+    }
+
+    @Override
+    public void onRenderedFirstFrame() {
+        Log.v("MediaSession", "onRenderedFirstFrame: state=[" + state + "]");
+        if (state == PAUSED) {
+            client.sendPauseRequest();
+        }
+    }
+
+    private void startKeepAlive() {
+        if (state == PREPARED) {
+            keepAliveEnabled = true;
+
+            if (timeout > 0) {
+                keepAliveRunnable();
+            }
+        }
+    }
+
+    private void stopKeepAlive() {
+        if (state > PREPARED && state < STOPPED) {
+            if (!executor.isShutdown()) {
+                executor.shutdown();
+            }
+
+            keepAliveEnabled = false;
+            state = STOPPED;
+        }
+    }
+
+    private void keepAliveRunnable() {
+        executor.execute(new Runnable() {
+
+            @Override
+            public void run() {
+                try {
+
+                    while (keepAliveEnabled) {
+
+                        try {
+
+                            while (!Thread.currentThread().isInterrupted() && keepAliveEnabled) {
+                                Thread.sleep(timeout / 2);
+                                client.sendKeepAlive();
+                            }
+
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
+                        }
+                    }
+
+                } finally {
+                    if (!executor.isShutdown()) {
+                        executor.shutdown();
+                    }
+                }
+            }
+        });
+    }
+
+
+    public static class Builder {
+        Uri uri;
+        Client client;
+        String username;
+        String password;
+
+        public final Builder client(Client client) {
+            if (client == null) throw new NullPointerException("client is null");
+
+            buildCredentialsByUri(client.uri());
+
+            this.client = client;
+            return this;
+        }
+
+        public final MediaSession build() {
+            if (client == null) throw new IllegalStateException("client is null");
+
+            return new MediaSession(this);
+        }
+
+        private void buildCredentialsByUri(Uri uri) {
+            if (uri == null) throw new NullPointerException("uri is null");
+
+            this.uri = Uri.parse(uri.getScheme() + "://" + uri.getHost() + ":" + uri.getPort() +
+                    uri.getPath() + ((uri.getQuery() != null) ? "?" + uri.getQuery() : ""));
+
+            if (uri.getUserInfo() != null) {
+                String[] values = uri.getUserInfo().split(":");
+
+                this.username = values[0];
+                this.password = values[1];
+
+                if (username == null) throw new IllegalStateException("username is null");
+                if (password == null) throw new IllegalStateException("password is null");
+            }
+
+            this.uri = uri;
```

Comment:
@tresvecesseis Why do you assign to `this.uri` on line 459 when on line 472 it is replaced with `uri` parameter, I think the line 472 should be deleted and line 459 should have check for not filled port `((uri.getPort() > 0) ? ":" + uri.getPort() : "")`

Reply:
@jiriadamec The buildCredentialsByUri method was created in this pull request and apparently we had left some duplicate lines. Thanks.